src/cmap/auth/mongodb_oidc/callback_workflow.ts

```diff
@@ -0,0 +1,209 @@
+import { type Document, Binary, BSON } from 'bson';
+import { promisify } from 'util';
+
+import { MongoInvalidArgumentError } from '../../../error';
+import { ns } from '../../../utils';
+import type { Connection } from '../../connection';
+import type { MongoCredentials } from '../mongo_credentials';
+import type { OIDCMechanismServerStep1, OIDCRequestTokenResult } from '../mongodb_oidc';
+import { AuthMechanism } from '../providers';
+import { TokenEntryCache } from './token_entry_cache';
+import type { Workflow } from './workflow';
+
+/* 5 minutes in milliseconds */
+const TIMEOUT = 300000;
+
+/**
+ * OIDC implementation of a callback based workflow.
+ * @internal
+ */
+export class CallbackWorkflow implements Workflow {
+  cache: TokenEntryCache;
+
+  /**
+   * Instantiate the workflow
+   */
+  constructor() {
+    this.cache = new TokenEntryCache();
+  }
+
+  /**
+   * Get the document to add for speculative authentication. Is empty when
+   * callbacks are in play.
+   */
+  speculativeAuth(): Promise<Document> {
+    return Promise.resolve({});
+  }
+
+  /**
+   * Execute the workflow.
+   *
+   * Steps:
+   * - If an entry is in the cache
+   *   - If it is not expired
+   *     - Skip step one and use the entry to execute step two.
+   *   - If it is expired
+   *     - If the refresh callback exists
+   *       - remove expired entry from cache
+   *       - call the refresh callback.
+   *       - put the new entry in the cache.
+   *       - execute step two.
+   *     - If the refresh callback does not exist.
+   *       - remove expired entry from cache
+   *       - call the request callback.
+   *       - put the new entry in the cache.
+   *       - execute step two.
+   * - If no entry is in the cache.
+   *   - execute step one.
+   *   - call the refresh callback.
+   *   - put the new entry in the cache.
+   *   - execute step two.
+   */
+  async execute(connection: Connection, credentials: MongoCredentials): Promise<Document> {
+    const entry = this.cache.getEntry(connection.address, credentials.username);
+    if (entry) {
+      // Check if the entry is not expired.
+      if (entry.isValid()) {
+        // Skip step one and execute the step two saslContinue.
+        try {
+          const result = await finishAuth(entry.tokenResult, undefined, connection, credentials);
+          return result;
+        } catch (error) {
+          // If authentication errors when using a cached token we remove it from
+          // the cache.
+          this.cache.deleteEntry(connection.address, credentials.username);
+          return Promise.reject(error);
```

Comment:
```suggestion
          throw error;
```
any reason we can't directly throw the error, since we're in an async function?

Reply:
Updated.